OCPBUGS-59657: allow default network -> localnet on the same node for any localnet subnet

ricky-rav opened the pull request:
Previous work (https://github.com/ovn-kubernetes/ovn-kubernetes/pull/5041) allowed default network -> localnet on the same node when host network and localnet shared the same subnet.

Let's extend this to allow the localnet to use any subnet.

The localnet being on a different network, we need an external router to connect host network and localnet:
```
       ┌────────────────────────────────────────┐
       │ external router                        │
       │                                        │
       │                                        │
       ┌──────────────┐                         │
       │ eth0.100     │                         │
       │ 172.20.0.5   │                         │
       │ eth0         │  net.ipv4.ip_forward=1  │
       │ 172.18.0.5   │                         │
       └──────────────┼─────────────────────────┘
           │
           │
           │
┌──────┌────────────┐──────────────────────────────────────────────────────────┐
│      │ eth0       │                                                          │
│      └───┬────────┘                                                          │
│          │                                                                   │
│    ┌────────────┐     ┌───────┐       ┌────────────────────────┐             │
│    │            │─────┤       ┼───────┼  pod in default network│             │
│    │  br-ex     │     │ br-int│       │  eth0                  │             │
│    │  172.18.0.4│     │       │       │  10.244.0.4            │             │
│    │            ┼─────┤       │       └────────────────────────┘             │
│    └────────────┘     └────┬──┘                                              │
│                            │                                                 │
│                      ┌─────┴────────────┐                                    │
│                      │ net1             │                                    │
│                      │ 172.20.0.4       │                                    │
│                      │                  │     nad: "vlanID": 100             │
│                      │ localnet pod     │                                    │
│                      │                  │                                    │
│                      │ ip route add 172.18.0.0/16 via 172.20.0.5 dev net1    │
│                      └──────────────────┘                                    │
└──────────────────────────────────────────────────────────────────────────────┘
```
The traffic path for a packet from the default network pod to the localnet pod is:
<img width="1006" height="796" alt="simple_diagram_with_arrow" src="https://github.com/user-attachments/assets/971def00-d55a-4069-83bd-4353d0b4327e" />
And the reverse is true for the reverse path. 

The reply traffic, from the localnet pod to the default network, is challenging, because it incorrectly matches the flows at prio 100 table=1 that output reply traffic directly to OVN or the host.

We overcome that by making sure that the flow in table=0 at priority=50 that sends traffic to conntrack & table=1 is only applied to traffic whose next hop is OVN/host. To do so, we add a match against `dl_dst=$breth0`:
```
 cookie=0xdeff105, duration=719.037s, table=0, n_packets=109738, n_bytes=232311922, idle_age=0, priority=50,ip,dl_dst=02:42:ac:12:00:03 actions=ct(table=1,zone=64000,nat)
 cookie=0xdeff105, duration=719.037s, table=0, n_packets=33, n_bytes=3804, idle_age=344, priority=50,ipv6,dl_dst=02:42:ac:12:00:03 actions=ct(table=1,zone=64000,nat)
[...]
```

Traffic from localnet will match the NORMAL flow at prio=0 and will be dispatched to the external router:
```
cookie=0x0, duration=694.665s, table=0, n_packets=2071, n_bytes=1781604, idle_age=18, priority=0 actions=NORMAL
```

It's worth pointing out that reply traffic from localnet hits br-ex twice:
- from localnet pod to external router: packets won't hit conntrack / table=1 and will just match table=0 prio=0 NORMAL rule
- from external router to the default network pod: packets will hit the table=0 prio=50 flow because their next hop is breth0; we lookup conntrack and jump to table=1. The packets are then correctly dispatched to OVN/host.


**E2E tests**
The setup that the E2E tests create makes use of an external container acting as a router, with eth0 configured with an IP on the host subnet and a VLAN interface eth0.100 configured with an IP in an external network (172.20.0.0/16) that the localnet uses.

```
                           +-----------------------+
                           |     Kubernetes Node   |
                           |       ovn-worker2     |
                           |                       |
   podA (10.244.1.10/24)---+---[ br-int ]----------+--- podL (172.20.0.4/16, net1)
   (default net)           |                       |   (localnet, VLAN 100)
                           |                       |
                           |       [ br-ex ]       |
                           |        172.18.0.2     |
                           +-----------|------------+
                                       |
                                 host network
                                 172.18.0.0/16
                                       |
                                      |
                          +------------------------+
                          |     ext. router     |
                          |                             |
                          |  eth0: 172.18.x.x      |
                          |  eth0.100: 172.20.0.5  |
                          +------------------------+

Packet path (request traffic podA → podL):
  podA (10.244.1.10) 
       → br-int 
          → br-ex (172.18.0.2) 
              → eth0 (external-container, 172.18.x.x) 
                → eth0.100 (external-container, 172.20.0.6, VLAN 100 GW) 
                  → eth0 (external-container, 172.18.x.x) 
                     → br-ex (172.18.0.2) 
                       → br-int 
                          → podL

The reply packet follows the reverse path.

```

In the e2e tests, I added:
- host network -> localnet, when the localnet uses IPs in the host subnet (missing from previous implementation)
- default network -> localnet, when the localnet is on an external subnet + VLAN
- host network -> localnet, when the localnet is on an external subnet + VLAN
- localnet  -> host network , when the localnet is on an external subnet + VLAN





<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Bridge MAC–based matching for external/related IPv4/IPv6 traffic so egress flows align with bridge routing and connection tracking.

* **Tests**
  * Expanded multihoming E2E: external‑router scenarios, VLAN/localnet and dual‑stack cases, route injection/cleanup framework, IP allocation from host/external subnets, interface‑aware checks, IP generation/normalization utilities.

* **Bug Fixes**
  * Improved test error handling and retry robustness (more ICMP attempts; clearer attachment status errors).

* **Chores**
  * CI: updated multihoming test job gateway-mode from local to shared.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

ricky-rav commented on go-controller/pkg/node/bridgeconfig/bridgeflows.go:
```diff
@@ -273,65 +273,118 @@ func (b *BridgeConfiguration) flowsForDefaultBridge(extraIPs []net.IP) ([]string
-					nodetypes.DefaultOpenFlowCookie, match_vlan, nodetypes.CtMarkHost, strip_vlan, ofPortHost))
+					nodetypes.DefaultOpenFlowCookie, match_vlan, bridgeMacAddress, nodetypes.CtMarkHost, strip_vlan, ofPortHost))
+
+			// table 1, established and related connections in zone 64000 with ct_mark CtMarkHost are applied a NORMAL action.
+			// This allows reply traffic that is routed through an external router to be sent to its learned next hop rather
+			// than directly to OVN.
+			dftFlows = append(dftFlows,
+				fmt.Sprintf("cookie=%s, priority=90, table=1, %s ip6, ct_state=+trk+est, ct_mark=%s, "+
+					"actions=%sNORMAL",
+					nodetypes.DefaultOpenFlowCookie, match_vlan, nodetypes.CtMarkHost, strip_vlan))
 
+			dftFlows = append(dftFlows,
+				fmt.Sprintf("cookie=%s, priority=90, table=1, %s ip6, ct_state=+trk+rel, ct_mark=%s, "+
+					"actions=%sNORMAL",
+					nodetypes.DefaultOpenFlowCookie, match_vlan, nodetypes.CtMarkHost, strip_vlan))
```
They are synonyms in openflow

maiqueb reviewed (changes requested):
I agree with the overall approach, but I'm leaving some comments. 

Most of them are opinionated nits meant to improve the code readability.

ricky-rav commented on go-controller/pkg/node/bridgeconfig/bridgeflows.go:
```diff
@@ -295,6 +297,17 @@ func (b *BridgeConfiguration) flowsForDefaultBridge(extraIPs []net.IP) ([]string
+				// This allows reply traffic that is routed through an external router to be sent to its learned next hop rather
+				// than directly to OVN. All traffic directly destined to OVN will have been matched at priority 100.
+				dftFlows = append(dftFlows,
+					fmt.Sprintf("cookie=%s, priority=90, table=1, ip, ct_state=+trk+est, ct_mark=%s, "+
+						"actions=%s", nodetypes.DefaultOpenFlowCookie, netConfig.MasqCTMark, actionsForExternallyRoutedTraffic))
+
+				dftFlows = append(dftFlows,
+					fmt.Sprintf("cookie=%s, priority=90, table=1,  ip, ct_state=+trk+rel, ct_mark=%s, "+
+						"actions=%s", nodetypes.DefaultOpenFlowCookie, netConfig.MasqCTMark, actionsForExternallyRoutedTraffic))
```
yes, that's the only difference. I'll add `tracked` to the comments I added around this code to make it crystal clear. I also wondered about it when first looking at this... :)

ricky-rav commented on test/e2e/multihoming.go:
```diff
@@ -276,39 +281,14 @@ var _ = Describe("Multi Homing", feature.MultiHoming, func() {
-			//              +------------------------------------------+
-			//
-			// We test podA when it sits on top of the overlay network, as depicted above, and
-			// when it is host-networked.
```
Here I moved the diagram further below. Let me put it in the final position from the start, it'll make more sense.

ricky-rav commented on test/e2e/multihoming.go:
```diff
@@ -2343,3 +2460,274 @@ func generateIPsFromSubnets(subnets []string, offset int) ([]string, error) {
+	return nil
+}
+
+func createExternalRouter(providerCtx infraapi.Context, cs clientset.Interface, f *framework.Framework, vlanID, ipOffset int) string {
```
You're right, I'll make it all return errors instead.

ricky-rav commented on test/e2e/multihoming.go:
```diff
@@ -2343,3 +2460,274 @@ func generateIPsFromSubnets(subnets []string, offset int) ([]string, error) {
+	// - Enable IP forwarding.
+	// - sleep to keep the container running.
+	var commandBuilder strings.Builder
+	commandBuilder.WriteString(fmt.Sprintf("ip link add link eth0 name eth0.%d type vlan id %d; ", vlanID, vlanID))
```
I create the external container for every test that needs it, so maybe it's not necessary to make the above command idempotent?

Otherwise I would just do:

```suggestion
	commandBuilder.WriteString(fmt.Sprintf("ip link show eth0.%d >/dev/null 2>&1 || ip link add link eth0 name eth0.%d type vlan id %d; ", vlanID, vlanID, vlanID))
```

ricky-rav commented on test/e2e/multihoming.go:
```diff
@@ -366,25 +344,36 @@ var _ = Describe("Multi Homing", feature.MultiHoming, func() {
+					serverPodConfig.usesExternalRouter && serverPodConfig.attachments != nil {
+					By("instantiating the external container")
+					externalRouterName := createExternalRouter(providerCtx, cs, f, netConfig.vlanID, externalRouterIPOffset)
+					Expect(externalRouterName).NotTo(BeEmpty(), "Failed to create external router: createExternalRouter returned empty router name")
```
Yeah, I got carried away with the reviews from coderabbit. I've removed that check.

ricky-rav commented on test/e2e/multihoming.go:
```diff
@@ -2343,3 +2460,274 @@ func generateIPsFromSubnets(subnets []string, offset int) ([]string, error) {
+
+	for _, ip := range gwV6IPs {
+		// filter out link local addresses
+		if addr, err := netip.ParseAddr(ip); err == nil && !addr.IsLinkLocalUnicast() {
+			gwIPV6 = ip
+			break
+		}
```
ok, I'll remove the `break`

maiqueb replied:
FWIW I still see the break

ricky-rav replied:
yeah, I changed my mind afterwards, sorry :D 
Here I'm just retrieving the IP addresses on the external container interface to use as gateways. I'm setting the IPs myself, for now no need to account for multiple IPv6 addresses there :)

ricky-rav commented on test/e2e/multihoming.go:
```diff
@@ -366,25 +344,36 @@ var _ = Describe("Multi Homing", feature.MultiHoming, func() {
 				By("instantiating the client pod")
 				kickstartPod(cs, clientPodConfig)
 
+				// Add external container that will act as external router for the localnet
+				if clientPodConfig.usesExternalRouter && clientPodConfig.attachments != nil ||
+					serverPodConfig.usesExternalRouter && serverPodConfig.attachments != nil {
+					By("instantiating the external container")
+					externalRouterName := createExternalRouter(providerCtx, cs, f, netConfig.vlanID, externalRouterIPOffset)
+					Expect(externalRouterName).NotTo(BeEmpty(), "Failed to create external router: createExternalRouter returned empty router name")
+
+					By("injecting routes via the external container")
+					injectStaticRoutesViaExternalContainer(f, cs, clientPodConfig, serverPodConfig, externalRouterName, netConfig.vlanID)
+				}
```
Here with a wrapper that calls  `createExternalRouter` and `injectStaticRoutesViaExternalContainer`  I would only save the bottom two lines :) Is this what you had in mind?

maiqueb replied:
I am just trying to establish the notion that a test doesn't need to have all the details shoved in the readers face. 

IMHO, if proper abstractions are used, we would read the high level test, understand what's being done, and be able to drill down on the details when needed / desired.

Of course it is a lot easier said than done, and "proper abstractions" mean probably a gazillion different things :) 

Let's keep it as is if you think the cost offsets the gains.

maiqueb commented on test/e2e/util.go:
```diff
@@ -823,6 +823,66 @@ func assertACLLogs(targetNodeName string, policyNameRegex string, expectedACLVer
 
+// getExternalContainerInterfaceIPsOnNetwork returns the IPv4 and IPv6 addresses (if any)
+// of the given external container on the specified provider network.
+func getExternalContainerInterfaceIPsOnNetwork(containerName, networkName string) (string, string, error) {
```
> I don't think I should pretend I don't know which one is which

I don't doubt *you* know which one is which. But you're not the only person reading the code. 

Your call, I won't push further.

maiqueb commented on test/e2e/multihoming.go:
```diff
@@ -2343,3 +2460,274 @@ func generateIPsFromSubnets(subnets []string, offset int) ([]string, error) {
+	var commandBuilder strings.Builder
+	commandBuilder.WriteString(fmt.Sprintf("ip link add link eth0 name eth0.%d type vlan id %d; ", vlanID, vlanID))
+	for _, ip := range routerIPs {
+		commandBuilder.WriteString(fmt.Sprintf("ip addr add %s dev eth0.%d; ", ip, vlanID))
```
resolving like the previous comment; idempotency here is overkill.

maiqueb commented on test/e2e/multihoming.go:
```diff
@@ -2198,53 +2524,526 @@ func computeIPWithOffset(baseAddr string, increment int) (string, error) {
-	node, err := cs.CoreV1().Nodes().Get(context.Background(), nodeName, metav1.GetOptions{})
+// buildRouteToHostSubnetViaExternalContainer returns ip route add commands to reach the host subnets via the provided gateway IPs
+func buildRouteToHostSubnetViaExternalContainer(cs clientset.Interface, f *framework.Framework, nodeName string, gwV4, gwV6, interfaceName string) ([]string, error) {
+	cmdTemplateV4 := "ip -4 route replace %s via " + gwV4 + " dev " + interfaceName
+	cmdTemplateV6 := "ip -6 route replace %s via " + gwV6 + " dev " + interfaceName
```
I think using `fmt.Sprintf` is both more readable and more efficient (since you don't have to allocate and append strings 3 times. 

Your call though - this isn't production code, so it's not that important.

maiqueb commented on test/e2e/multihoming.go:
```diff
@@ -2343,3 +2460,274 @@ func generateIPsFromSubnets(subnets []string, offset int) ([]string, error) {
+
+		framework.Logf("Shared gateway mode: injecting Linux kernel routes for host-networked pods")
+		injectRoutesWithCommandBuilder(f, cs, externalContainerName, hostRoutingTableCommandBuilder{})
+	}
```
Nice, tks.

maiqueb commented on test/e2e/multihoming.go:
```diff
@@ -2343,3 +2460,274 @@ func generateIPsFromSubnets(subnets []string, offset int) ([]string, error) {
+			ovnkubePods, err = cs.CoreV1().Pods(ovnKubernetesNamespace).List(context.TODO(), metav1.ListOptions{LabelSelector: "ovn-db-pod=true"})
+			framework.ExpectNoError(err, "failed to list DB leader pods")
+		}
+	}
```
lovely !

maiqueb commented on test/e2e/multihoming.go:
```diff
@@ -2343,3 +2460,274 @@ func generateIPsFromSubnets(subnets []string, offset int) ([]string, error) {
+	v4gateway, v6gateway, err := getExternalContainerInterfaceIPsOnNetwork(externalContainerName, primaryProviderNetwork.Name())
+	framework.ExpectNoError(err, "failed to get external container IPs on provider network")
+
+	ovnKubernetesNamespace := deploymentconfig.Get().OVNKubernetesNamespace()
```
oh that's even better ! Nice ! Please resolve this thread once you push that improvement.

Thank yoU!

maiqueb commented on go-controller/pkg/node/bridgeconfig/bridgeflows.go:
```diff
@@ -273,65 +273,120 @@ func (b *BridgeConfiguration) flowsForDefaultBridge(extraIPs []net.IP) ([]string
 				dftFlows = append(dftFlows,
-					fmt.Sprintf("cookie=%s, priority=100, table=1, ip, ct_state=+trk+est, ct_mark=%s, "+
-						"actions=%s", nodetypes.DefaultOpenFlowCookie, netConfig.MasqCTMark, actions))
+					fmt.Sprintf("cookie=%s, priority=100, table=1, dl_dst=%s, ip, ct_state=+trk+est, ct_mark=%s, "+
```
do we want this " " after the bridge's MAC here ? Is it harmful in any way ? 

Does somehow openflow become "cheaper" if we remove it ?

maiqueb commented on test/e2e/multihoming.go:
```diff
@@ -444,18 +538,248 @@ var _ = Describe("Multi Homing", feature.MultiHoming, func() {
+			//   podA (10.244.1.10)
+			//     → br-int
+			//       → br-ex (172.18.0.2, SNAT to node IP)
+			//         → eth0 (external router, 172.18.0.55)
```
Is the IP here supposed to be `172.20.0.55` instead ? I'm a bit confused by the .55 within the 172.18.0.0 subnet.

ricky-rav replied:
I've removed that, it's not important for the diagram

ricky-rav commented on test/e2e/util.go:
```diff
@@ -823,6 +824,66 @@ func assertACLLogs(targetNodeName string, policyNameRegex string, expectedACLVer
+// for manually-configured interfaces like VLAN interfaces.
+func getExternalContainerInterfaceIPs(containerName, ifaceName string) ([]string, []string, error) {
+	container := infraapi.ExternalContainer{Name: containerName}
+
+	// Replicates the relevant fields from the json output by "ip -j addr show"
+	type addrInfo struct {
+		Family string `json:"family"`
+		Local  string `json:"local"`
+	}
+	type ipAddrJSON struct {
+		AddrInfo []addrInfo `json:"addr_info"`
+	}
+
+	cmd := []string{"sh", "-c", fmt.Sprintf("ip -j addr show dev %s", ifaceName)}
+	out, err := infraprovider.Get().ExecExternalContainerCommand(container, cmd)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to exec on container %q: %w", containerName, err)
+	}
+	var parsed []ipAddrJSON
+	if err := json.Unmarshal([]byte(out), &parsed); err != nil {
+		return nil, nil, fmt.Errorf("failed to parse ip -j output: %w", err)
+	}
+
+	var v4, v6 []string
+	for _, entry := range parsed {
+		for _, ai := range entry.AddrInfo {
+			if ai.Local == "" {
+				continue
+			}
+			switch ai.Family {
+			case "inet":
+				v4 = append(v4, ai.Local)
+			case "inet6":
+				v6 = append(v6, ai.Local)
+			}
+		}
+	}
+
+	return v4, v6, nil
+}
```
agreed

ricky-rav commented on test/e2e/multihoming.go:
```diff
@@ -36,8 +36,58 @@ import (
+// If the pod is host-networked, it returns eth0.
+// If the pod has attachments, it finds the interface for the specified network
+// If the pod has no attachments, it returns the default network interface
+func getNetworkInterfaceName(pod *v1.Pod, podConfig podConfiguration, netConfigName, namespace string) (string, error) {
```
yup

maiqueb commented on test/e2e/multihoming.go:
```diff
@@ -2198,53 +2524,524 @@ func computeIPWithOffset(baseAddr string, increment int) (string, error) {
-		nodeAddresses = append(nodeAddresses, nodeIfAddr.IPv6)
+
+	var (
+		IPsToRequest []string
```
pedantic nit: not sure this counts as breaking the variable naming convention (start w/ lowercase). 

```suggestion
		ipsToRequest []string
```

maiqueb commented on test/e2e/multihoming.go:
```diff
@@ -2198,53 +2524,524 @@ func computeIPWithOffset(baseAddr string, increment int) (string, error) {
+	return nil
+}
+
+// routeCommand represents a route add/delete command for a specific gateway mode
+type routeCommand struct {
+	addCmd    []string
+	deleteCmd []string
+	logMsg    string
+	target    string // what we're adding the route to (logical router name or node name)
+}
+
+// routeCommandBuilder defines the interface for building gateway-mode-specific route commands
+type routeCommandBuilder interface {
+	buildRouteCommand(nodeName, cidr, nextHop string) routeCommand
+}
+
+// ovnLogicalRouterCommandBuilder builds commands for OVN logical routes (shared gateway mode)
+type ovnLogicalRouterCommandBuilder struct{}
+
+func (b ovnLogicalRouterCommandBuilder) buildRouteCommand(nodeName, cidr, nextHop string) routeCommand {
+	logicalRouterName := "GR_" + nodeName
+	return routeCommand{
+		addCmd:    []string{"ovn-nbctl", "--may-exist", "--", "lr-route-add", logicalRouterName, cidr, nextHop},
+		deleteCmd: []string{"ovn-nbctl", "--if-exists", "--", "lr-route-del", logicalRouterName, cidr},
+		logMsg:    fmt.Sprintf("OVN logical router route %s via %s to %s", cidr, nextHop, logicalRouterName),
+		target:    logicalRouterName,
+	}
+}
```
nit: maybe you could offload this to a separate file / package.

maiqueb commented on test/e2e/multihoming.go:
```diff
@@ -339,27 +387,46 @@ var _ = Describe("Multi Homing", feature.MultiHoming, func() {
-				} else {
+				if serverPodConfig.attachments != nil {
 					serverIPs, err = podIPsForAttachment(cs, serverPod.Namespace, serverPod.Name, netConfig.name)
-
+				} else {
+					serverIPs, err = podIPsFromStatus(cs, serverPodConfig.namespace, serverPodConfig.name)
```
yes, I understand that; the point I'm making is I think you will also find the IPs in the network status. 

After all, that's an encapsulated version the CNI ADD response. The IPs should be there. 

Anyway, you don't need to change this. That can be sorted out on a separate PR.

ricky-rav commented:
After talking to @trozet about his PR https://github.com/ovn-kubernetes/ovn-kubernetes/pull/5529, we decided that the correct way forward is his approach to rely on a modified table=0 prio=50 rule that sends to conntrack and table=1 only traffic whose next hop is OVN/host. Indeed, table=1 is only intended for such traffic. Using it for localnet traffic destined to an external router wasn't correct, since we're not dispatching it to OVN/host, but to eth0.  I've updated the PR description to reflect that, removed my openflow commits and added Tim's.

trozet reviewed (dismissed):
/lgtm

but I did not have time to review line by line for the tests. Will defer to @tssurya to also lgtm and merge

maiqueb commented on test/e2e/multihoming_external_router_utils.go:
```diff
@@ -0,0 +1,429 @@
+	//
+	// 3) in the host routing table, if the cluster is in local gateway mode, we need a route
+	//    that tells the host to route traffic to the localnet subnet via the external container IPs.
+	if clientPodConfig.usesExternalRouter && clientPodConfig.attachments != nil {
+		if err := injectRouteViaExternalContainerIntoPod(f, cs, clientPodConfig, clientInterface, externalContainerName, vlanID); err != nil {
+			return fmt.Errorf("failed to add route to client pod %s/%s: %w", clientPodConfig.namespace, clientPodConfig.name, err)
+		}
+	}
+
+	if serverPodConfig.usesExternalRouter && serverPodConfig.attachments != nil {
+		if err := injectRouteViaExternalContainerIntoPod(f, cs, serverPodConfig, serverInterface, externalContainerName, vlanID); err != nil {
+			return fmt.Errorf("failed to add route to server pod %s/%s: %w", serverPodConfig.namespace, serverPodConfig.name, err)
+		}
+	}
```
yeah, this one is legit.

ricky-rav commented on test/e2e/multihoming_external_router_utils.go:
```diff
@@ -0,0 +1,429 @@
+		return fmt.Errorf("failed to get primary network: %w", err)
+	}
+
+	v4gateway, v6gateway, err := getExternalContainerInterfaceIPsOnNetwork(externalContainerName, primaryProviderNetwork.Name())
+	if err != nil {
+		return fmt.Errorf("failed to get external container interface IPs on provider network: %w", err)
+	}
+
+	// Build list of routes to inject
+	routes, err := buildRouteList(f.ClientSet, v4gateway, v6gateway)
+	if err != nil {
+		return err
+	}
```
OK

ricky-rav commented:
> @ricky-rav I see the virt lane is broken: https://github.com/ovn-kubernetes/ovn-kubernetes/actions/runs/17320600834/job/49175066594?pr=5480
> 
> Maybe you broke something with this fix ? We need to ensure those tests do no regress.

Sure, I'll run it again. For now I don't see any relation:
- one test failed because `killall iperf3` failed:
```
2025-08-29T11:18:55.6746975Z   Unexpected error:
2025-08-29T11:18:55.6747458Z       <*errors.errorString | 0xc00099c310>: 
2025-08-29T11:18:55.6748296Z       failed to run [killall iperf3] at VMI worker-spjvs, error: killall iperf3 failed
2025-08-29T11:18:55.6748898Z       {
2025-08-29T11:18:55.6750087Z           s: "failed to run [killall iperf3] at VMI worker-spjvs, error: killall iperf3 failed",
2025-08-29T11:18:55.6750705Z       }
2025-08-29T11:18:55.6751090Z   occurred[0m
```
- the other test reported that iperf3 had its connect reset after initially working:
```
2025-08-29T11:13:21.2292089Z [0mKubevirt Virtual Machines [38;5;243mwith user defined networks and persistent ips configured [0mshould keep ip [38;5;9m[1m[It] after live migration of VirtualMachine with interface binding for UDN with Primary/Layer2 with snat ingress[0m [38;5;204m[Feature:VirtualMachineSupport][0m
2025-08-29T11:13:21.2328184Z [38;5;243m/home/runner/work/ovn-kubernetes/ovn-kubernetes/test/e2e/kubevirt.go:2033[0m
2025-08-29T11:13:21.2328697Z 
2025-08-29T11:13:21.2329783Z   [38;5;9m[FAILED] worker-6ktwc: Check north/south traffic after VirtualMachine with interface binding for UDN live migration: /tmp/ingress_test_172.18.0.3_31488_iperf3.log
2025-08-29T11:13:21.2330712Z   Expected
2025-08-29T11:13:21.2331173Z       <string>: Connecting to host 172.18.0.3, port 31488
2025-08-29T11:13:21.2332092Z       [  6] local 172.18.0.7 port 51684 connected to 172.18.0.3 port 31488
2025-08-29T11:13:21.2332839Z       [ ID] Interval           Transfer     Bitrate         Retr  Cwnd
2025-08-29T11:13:21.2333623Z       [  6]   0.00-1.00   sec   532 MBytes  4.46 Gbits/sec   34   3.23 MBytes       
2025-08-29T11:13:21.2334426Z       [  6]   1.00-2.00   sec   538 MBytes  4.51 Gbits/sec    0   3.23 MBytes       
[...]
2025-08-29T11:13:21.2492417Z       [  6]  51.00-52.00  sec   316 MBytes  2.65 Gbits/sec    0   3.23 MBytes       
2025-08-29T11:13:21.2493285Z       [  6]  52.00-53.00  sec   302 MBytes  2.54 Gbits/sec    0   3.23 MBytes       
2025-08-29T11:13:21.2494157Z       [  6]  53.00-54.00  sec   319 MBytes  2.67 Gbits/sec    0   3.23 MBytes       
2025-08-29T11:13:21.2495327Z       [  6]  54.00-55.00  sec   299 MBytes  2.51 Gbits/sec    0   3.23 MBytes       
2025-08-29T11:13:21.2522022Z       iperf3: error - unable to write to stream socket: Connection reset by peer
2025-08-29T11:13:21.2522625Z       
2025-08-29T11:13:21.2522968Z   not to contain substring
2025-08-29T11:13:21.2524502Z       <string>: iperf3: error[0m
```

tssurya commented on test/e2e/multihoming.go:
```diff
@@ -369,8 +368,34 @@ var _ = Describe("Multi Homing", feature.MultiHoming, func() {
 					}, 2*time.Minute, 6*time.Second).Should(Succeed())
 				}
 			},
+			// We test here a setup with a localnet that uses IPs in the host subnet:
```
love the diagram

tssurya commented on test/e2e/multihoming.go:
```diff
@@ -369,8 +368,34 @@ var _ = Describe("Multi Homing", feature.MultiHoming, func() {
+			//                         |        172.18.0.2     |
+			//                         |           |           |
+			//                         |           |           |
+			//                         |   podL (172.18.0.4)   |
```
~what is podL?~
podLocalnet

tssurya commented on test/e2e/multihoming.go:
```diff
@@ -412,7 +437,7 @@ var _ = Describe("Multi Homing", feature.MultiHoming, func() {
 			),
 			Entry(
-				"can reach a host-networked pod on a different node",
+				"can reach a host-networked pod on a different node, when the localnet uses an IP in the host subnet",
```
is this still a client pod in default network as the source or the localnetpod to hostnetwork pod in default network?

tssurya commented on test/e2e/multihoming.go:
```diff
@@ -452,7 +452,51 @@ var _ = Describe("Multi Homing", feature.MultiHoming, func() {
+				Label("STORY", "SDN-5345"),
+			),
+			Entry(
+				"can be reached by a host-networked pod on a different node",
```
nit: this name and the ones above good to clarify host-neteworked pod on a different node to local net pod v/s localnet->host is the can reach is it?

so can be reached is host-networked pod to localnet pod
and can reach is localnet pod to host-networked pod

ricky-rav replied:
I had to play with the verb (`can be reached` vs `can reach`) in order to be under the same outer `Context("A single pod with an OVN-K secondary network", func() {`. 
I'll add a comment to clarify what we're testing

tssurya commented on test/e2e/multihoming.go:
```diff
@@ -2316,3 +2320,26 @@ func addIPRequestToPodConfig(cs clientset.Interface, podConfig *podConfiguration
 	return nil
 }
+
+func generateIPsFromSubnets(subnets []string, offset int) ([]string, error) {
```
nit: add doc string its too generic here with "generateIPsFromSubnet"

tssurya replied:
haha I went commit by commit :)

tssurya commented on test/e2e/multihoming.go:
```diff
@@ -33,9 +33,9 @@ import (
-	fromHostSubnet          = "host" // tells the test to generate IPs from the host subnet.
+	PolicyForAnnotation = "k8s.v1.cni.cncf.io/policy-for"
+	nodeHostnameKey     = "kubernetes.io/hostname"
+	fromHostSubnet      = "host" // tells the test to generate IPs from the host subnet.
```
nit: this change should be in the `E2E localnet: make IP request for localnet pod extensible` but its also showing up in `E2E localnet: Fix requirement on number of schedulable nodes`

tssurya commented on test/e2e/multihoming.go:
```diff
@@ -276,10 +320,11 @@ var _ = Describe("Multi Homing", feature.MultiHoming, func() {
+			clientPodName          = "client-pod"
+			clientIPOffset         = 100
+			serverIPOffset         = 102
+			externalRouterIPOffset = 55
```
nit: oh what are these IP offsets? maybe we need a var description comment

tssurya commented on test/e2e/multihoming.go:
```diff
@@ -525,6 +588,92 @@ var _ = Describe("Multi Homing", feature.MultiHoming, func() {
+			//     → br-int
+			//       → br-ex (172.18.0.2, SNAT to node IP)
+			//         → eth0 (external router, 172.18.x.x)
+			//           → eth0.10 (external router, 172.20.0.55)
```
how does the `ip r` output and `ip a` output of this external router look like?
I assume you added some static routes to back into br-ex for the destination 172.20.0.4?

tssurya commented on test/e2e/multihoming_external_router_utils.go:
```diff
@@ -0,0 +1,429 @@
+package e2e
```
I didn't have too much time to look at the utils all too closely, was this AI-assisted?
I trust @maiqueb reviewed it when he did the +1 
my only concern is whether any of the existing code's re-usability was considered for the utils like with the ICNI test external gateway - but I assume you wanted something more specific here

ricky-rav replied:
I used cursor ai to help me speed up the implementation of some individual bits, but I did review and adjust/fix all code at every iteration. :) 

Looking at external_gateway.go and e2e.go, I don't see what I could have reused actually... for instance there's `setupGatewayContainers`, but it doesn't coincide 100% with what I have to do here.

tssurya commented on test/e2e/multihoming_external_router_utils.go:
```diff
@@ -0,0 +1,429 @@
+
+// getOvnKubePodsForRouteInjection determines which pods to use for route injection based on the command builder type
+// and cluster configuration.
+func getOvnKubePodsForRouteInjection(f *framework.Framework, cs clientset.Interface, cmdBuilder routeCommandBuilder) (*v1.PodList, error) {
```
this and some other utils here... are we sure we can't reuse existing utils from e2e.go file?
we have plumbing for some of this but maybe not exactly specific to your test here.

ricky-rav replied:
yeah, I didn't find exactly what I needed.

tssurya commented on test/e2e/multihoming.go:
```diff
@@ -35,9 +32,56 @@ import (
+// If the pod is host-networked, it returns eth0.
+// If the pod has attachments, it finds the interface for the specified network
+// If the pod has no attachments, it returns the default network interface
+func getNetworkInterfaceName(pod *v1.Pod, podConfig podConfiguration, netConfigName string) (string, error) {
```
nit: should this also be in utils?

tssurya commented on test/e2e/multihoming.go:
```diff
@@ -338,27 +381,46 @@ var _ = Describe("Multi Homing", feature.MultiHoming, func() {
+					Expect(err).NotTo(HaveOccurred())
+
+					By("injecting routes via the external container")
+					err = injectStaticRoutesViaExternalContainer(f, cs, clientPodConfig, serverPodConfig,
```
I think this test can benefit from a step-by-step plain english explanation like here: https://github.com/ovn-kubernetes/ovn-kubernetes/blob/56d14a3074ba7df00a6b68dfe4ac60aff7fc624a/test/e2e/egressip.go#L865 but including details around the nbctl routes you add and why.
having the step by step comments can help the next contributor who needs to modify this test and understand why we did the setup this way etc

tssurya commented on test/e2e/multihoming.go:
```diff
@@ -674,6 +674,56 @@ var _ = Describe("Multi Homing", feature.MultiHoming, func() {
 				Label("BUG", "OCPBUGS-59657"),
 			),
+			Entry(
+				"can be reached by a client pod in the host network on a different node, when the localnet uses a VLAN and an external router",
```
nit: `can be reached by a client host-networked pod on a different node....`

tssurya commented on test/e2e/multihoming.go:
```diff
@@ -674,6 +674,56 @@ var _ = Describe("Multi Homing", feature.MultiHoming, func() {
+					usesExternalRouter:  true,
+				},
+				true, // scheduled on the same node
+				Label("BUG", "OCPBUGS-59657"),
```
what are these Label's used for? these are for known issues right? but since we have fixed this issue why are downstream bugs being linked? we should remove these

tssurya commented on test/e2e/multihoming.go:
```diff
@@ -724,6 +724,56 @@ var _ = Describe("Multi Homing", feature.MultiHoming, func() {
 				Label("BUG", "OCPBUGS-59657"),
 			),
+			Entry(
+				"can reach a client pod in the host network on a different node, when the localnet uses a VLAN and an external router",
```
also same here nit: client host networked pod makes it a bit more easier to read but its a nit

tssurya commented on test/e2e/multihoming_utils.go:
```diff
@@ -358,7 +358,7 @@ func pingServer(clientPodConfig podConfiguration, serverIP string, args ...strin
 		"--",
 		"ping",
-		"-c", "1", // send one ICMP echo request
+		"-c", "3", // send three ICMP echo requests
```
are all these tests pings only? why do I feel like curl is a better test :) but I assume this existed even before you did tests right?

ricky-rav replied:
I run curl first and then I send three pings :)

tssurya reviewed (approved):
just a few nits and suggestions.
otherwise looks good to me
I didn't fully understand the external server eth0 and eth0.100 interface bits so maybe when i see the ip a and ip r output things might make more sense on how it hairpins back

ricky-rav commented on .github/workflows/test.yml:
```diff
@@ -450,6 +450,7 @@ jobs:
           - {"target": "node-ip-mac-migration", "ha": "noHA", "gateway-mode": "shared", "ipfamily": "ipv4",      "disable-snat-multiple-gws": "SnatGW",   "second-bridge": "1br", "ic": "ic-single-node-zones"}
           - {"target": "compact-mode",      "ha": "noHA", "gateway-mode": "local",  "ipfamily": "ipv4",      "disable-snat-multiple-gws": "snatGW",   "second-bridge": "1br", "ic": "ic-disabled"}
           - {"target": "multi-homing",      "ha": "noHA", "gateway-mode": "local",  "ipfamily": "dualstack", "disable-snat-multiple-gws": "SnatGW",   "second-bridge": "1br", "ic": "ic-single-node-zones", "network-segmentation": "enable-network-segmentation"}
+          - {"target": "multi-homing",      "ha": "noHA", "gateway-mode": "shared",  "ipfamily": "dualstack", "disable-snat-multiple-gws": "SnatGW",   "second-bridge": "1br", "ic": "ic-single-node-zones", "network-segmentation": "enable-network-segmentation"}
```
Sure! I didn't want to eliminate an existing one, since I don't know the history behind the non-IC lanes. @maiqueb WDYT?

ricky-rav replied:
@tssurya I'll fix this once we get a reply from NVIDIA https://cloud-native.slack.com/archives/C08452HR8V6/p1756751023229769 
The helm lane was last modified by Flavio back in April...

trozet commented:
/retest-failed

tssurya commented:
We are testing this PR downstream just as pre-merge testing, I'll circle back here tomorrow and merge this if things are green in this CI run and the downstream run
